File: app/src/api/scheduling/settings/services/vehicle-category-service.ts
import { http } from '@/api/shared/http';
import { VehicleCategory } from '@/api/scheduling/settings/data-transfer-objects/vehicle-category';
import { VehicleCategoryCreate } from '@/api/scheduling/settings/data-transfer-objects/vehicle-category-create';
import { VehicleCategoryUpdate } from '@/api/scheduling/settings/data-transfer-objects/vehicle-category-update';

export class VehicleCategoryService {
    public async getVehicleCategories() {
        try {
            var res = await http.get<[{ id: string; name: string; description?: string }]>(
                '/settings/vehicle-category'
            );

            return res.data.map(vc => {
                var newOne = new VehicleCategory(vc.id, vc.name, vc.description);
                newOne.id = vc.id;

                return newOne;
            });
        } catch (e) {
            if (e.response.status == 404) {
                return [];
            } else {
                throw e;
            }
        }
    }

    public async createVehicleCategory(create: VehicleCategoryCreate) {
        var res = await http.post('/settings/vehicle-category', create);

        var vc = new VehicleCategory(res.data.name, res.data.description);
        vc.id = res.data.id;

        return vc;
    }

    public async updateVehicleCategory(update: VehicleCategoryUpdate) {
        var res = await http.put(`/settings/vehicle-category/${update.id}`, update);

        var vc = new VehicleCategory(res.data.name, res.data.description);
        vc.id = res.data.id;

        return vc;
    }

    public async deleteVehicleCategory(vehicleCategory: VehicleCategory) {
        await http.delete(`/settings/vehicle-category/${vehicleCategory.id}`);
    }
}

export const vehicleCategoryService = new VehicleCategoryService();

File: api/src/Domain/Scheduling/Settings/Services/VehicleCategoryService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DetailingArsenal.Domain.Users;

namespace DetailingArsenal.Domain.Settings {
    public interface IVehicleCategoryService : IService {
        Task<VehicleCategory> GetById(Guid id);
        Task<List<VehicleCategory>> GetByUser(User user);
        Task<VehicleCategory> Create(VehicleCategoryCreate create, User user);
        Task Update(VehicleCategory category, VehicleCategoryUpdate update);
        Task Delete(VehicleCategory category);
    }

    [DependencyInjection(RegisterAs = typeof(IVehicleCategoryService))]
    public class VehicleCategoryService : IVehicleCategoryService {
        private VehicleCategoryNameUniqueSpecification uniqueNameSpec;
        private VehicleCategoryNotInUseSpecification notInUseSpec;
        private IVehicleCategoryRepo repo;

        public VehicleCategoryService(VehicleCategoryNameUniqueSpecification uniqueNameSpec, VehicleCategoryNotInUseSpecification notInUseSpec, IVehicleCategoryRepo repo) {
            this.uniqueNameSpec = uniqueNameSpec;
            this.notInUseSpec = notInUseSpec;
            this.repo = repo;
        }

        public async Task<VehicleCategory> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<VehicleCategory>> GetByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<VehicleCategory> Create(VehicleCategoryCreate create, User user) {
            var cat = VehicleCategory.Create(create, user);

            await uniqueNameSpec.CheckAndThrow(cat);
            await repo.Add(cat);

            return cat;
        }

        public async Task Update(VehicleCategory category, VehicleCategoryUpdate update) {
            category.Name = update.Name;
            category.Description = update.Description;

            await uniqueNameSpec.CheckAndThrow(category);
            await repo.Update(category);
        }

        public async Task Delete(VehicleCategory category) {
            await notInUseSpec.CheckAndThrow(category);
            await repo.Delete(category);
        }
    }
}